core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -1498,28 +1498,44 @@ class UnifiedLog(@volatile var logStartOffset: Long,
     producerStateManager.takeSnapshot()
     updateHighWatermarkWithLogEndOffset()
     // Schedule an asynchronous flush of the old segment
-    scheduler.schedule("flush-log", () => flush(newSegment.baseOffset))
+    scheduler.schedule("flush-log", () => flushUptoOffsetExclusive(newSegment.baseOffset))
     newSegment
   }
 
   /**
    * Flush all local log segments
+   *
+   * @param forceFlushActiveSegment should be true during a clean shutdown, and false otherwise. The reason is that
+   * we have to pass logEndOffset + 1 to the `localLog.flush(offset: Long): Unit` function to flush empty
+   * active segments, which is important to make sure we persist the active segment file during shutdown, particularly
+   * when it's empty.
    */
-  def flush(): Unit = flush(logEndOffset)
+  def flush(forceFlushActiveSegment: Boolean): Unit = flush(logEndOffset, forceFlushActiveSegment)
```

Comment:
The addition of this flag seems to be an optimization so that we are not forced to flush newly created (empty) segments. Do you have a sense for how valuable this optimization is? The additional noise it adds to the log API is a bit unfortunate, so it would be helpful to understand what we're getting in return.

Reply:
I will run a benchmark next week.

Reply:
@hachikuji The PR description needs to be updated, but the main take away from this PR is that it helps fix a durability corner case during clean shutdown. Before this PR, an empty active segment doesn't get flushed to disk during clean shutdown. This PR fixes it, and after this PR, it gets flushed. Without this fix, the loss of an empty active segment file on disk can cause the replica to lose track of logEndOffset.

Alongside the above fix, the PR also prevents some unwanted logging side effect in the recovery code path in LogLoader.

I agree that the noise to the flush API signature is unfortunate. I'm not sure if we can refactor to introduce a better API that can avoid the noise. But given that this PR fixes a durability case, the PR seems useful overall.

Reply:
@kowshik What I am suggesting is that we could do the inclusive flush in all cases in order to simplify the API if the performance benefit is minimal. In addition to the simpler API, it would fix the problem of losing track of the log end offset in the general case (not just for clean shutdown).

Reply:
IIUC that would increase the # of fsyncs that we'd typically do during segment rolls from 1 to 2. I'm not sure if we should make that change in this PR.

Reply:
@kowshik this is exactly what I wanted to test. I will start two clusters and see the perf difference between fsync twice and once during segment rolls. Will post the results here.

Reply:
Fsync is expensive. If we are introducing an extra fsync to disk during segment rolls, this could need extensive testing if we go down this route. cc @junrao

Reply:
If we have 50MB/s produce input in a three-broker cluster and segment.size is 100MB, it is on average 1 extra fsync every 2 seconds. Given the extra fsync is to sync an empty file, it may not be a big deal.